File: temoto_nl_interface/actions/find_coordinates/src/getcoord_py/getcoord_pathfind_return.py
# getcoord_pathfind_return.py

import json
import heapq
import numpy as np

def get(object_map, resolution, origin, items_data, robot_coords, assistant_reply):
    # Parse assistant_reply to get target_id
    assistant_data = json.loads(assistant_reply)
    target_id = assistant_data.get('target_id')

    # Find the target item coordinates in items_data
    item_coords = None
    for item_class, items_list in items_data.get('items', {}).items():
        for item in items_list:
            if item.get('id') == target_id:
                item_coords = item.get('coordinates')
                break
        if item_coords is not None:
            break

    if item_coords is None:
        assistant_data['success'] = "false"
        assistant_data['error'] = "Target ID not found in items_data"
        assistant_data['message'] = "The specified target could not be found."
        return json.dumps(assistant_data)

    # Get map dimensions
    map_height, map_width = object_map.shape[:2]  # Grayscale image has shape (height, width)

    # Extract origin coordinates (only x and y)
    origin_x, origin_y = origin[:2]  # Assuming origin is a list or tuple

    # Convert robot_coords and item_coords to map indices
    def world_to_map(x, y):
        col = int((x - origin_x) / resolution)
        row = map_height - int((y - origin_y) / resolution) - 1
        return (row, col)

    robot_x = robot_coords.get('x')
    robot_y = robot_coords.get('y')
    robot_position = world_to_map(robot_x, robot_y)

    item_x = item_coords.get('x')
    item_y = item_coords.get('y')
    item_position = world_to_map(item_x, item_y)

    # Check if indices are within map bounds
    def is_within_bounds(position):
        row, col = position
        return 0 <= row < map_height and 0 <= col < map_width

    if not is_within_bounds(robot_position):
        assistant_data['success'] = "false"
        assistant_data['error'] = "Robot coordinates out of map bounds"
        assistant_data['message'] = "The robot's position is outside the map boundaries."
        return json.dumps(assistant_data)

    if not is_within_bounds(item_position):
        assistant_data['success'] = "false"
        assistant_data['error'] = "Item coordinates out of map bounds"
        assistant_data['message'] = "The item's position is outside the map boundaries."
        return json.dumps(assistant_data)

    # Proceed with pathfinding
    class Node:
        def __init__(self, position, parent=None):
            self.position = position  # (row, col)
            self.parent = parent
            self.g = 0  # Cost from start to current node
            self.h = 0  # Estimated cost from current to goal
            self.f = 0  # Total cost

        def __lt__(self, other):
            return self.f < other.f

    def heuristic(a, b):
        # Use Euclidean distance as the heuristic
        return ((a[0] - b[0])**2 + (a[1] - b[1])**2) ** 0.5

    def is_walkable(position):
        row, col = position
        # Check if the position is within bounds and is a walkable (white) area
        if is_within_bounds(position):
            pixel = object_map[row][col]
            if object_map.ndim == 3:
                # For color images, check if all channels are 255
                return np.all(pixel == 255)
            else:
                # For grayscale images, check if the pixel is 255
                return pixel == 255
        return False

    # Initialize the open and closed lists
    open_list = []
    closed_set = set()

    start_node = Node(robot_position)
    start_node.h = heuristic(robot_position, item_position)
    start_node.f = start_node.h

    min_distance = start_node.h
    best_node = start_node

    heapq.heappush(open_list, start_node)

    # Define movement directions (up, down, left, right)
    movements = [(-1, 0), (1, 0), (0, -1), (0, 1)]

    while open_list:
        current_node = heapq.heappop(open_list)
        current_position = current_node.position

        if current_position in closed_set:
            continue

        closed_set.add(current_position)

        # Update the best_node if closer to the item
        current_distance = heuristic(current_position, item_position)
        if current_distance < min_distance:
            min_distance = current_distance
            best_node = current_node

        # If we reach the item position (unlikely since it's unreachable), we can break
        if current_position == item_position:
            best_node = current_node
            break

        # Explore neighbors
        for move in movements:
            neighbor_row = current_position[0] + move[0]
            neighbor_col = current_position[1] + move[1]
            neighbor_position = (neighbor_row, neighbor_col)

            if not is_walkable(neighbor_position) or neighbor_position in closed_set:
                continue

            neighbor_node = Node(neighbor_position, current_node)
            neighbor_node.g = current_node.g + 1
            neighbor_node.h = heuristic(neighbor_position, item_position)
            neighbor_node.f = neighbor_node.g + neighbor_node.h

            heapq.heappush(open_list, neighbor_node)

    # **Return Map Coordinates Instead of World Coordinates**
    final_coords = {
        'x': best_node.position[1],  # col index as x-coordinate
        'y': best_node.position[0]   # row index as y-coordinate
    }

    # Update assistant_reply with the new map coordinates
    assistant_data['coordinates'] = {
        'x': final_coords['x'],
        'y': final_coords['y']
    }

    # Ensure success is true and error is none
    assistant_data['success'] = "true"
    assistant_data['error'] = "none"

    # Convert back to string
    updated_assistant_reply = json.dumps(assistant_data)

    return updated_assistant_reply

File: temoto_nl_interface/actions/find_coordinates/src/get_coordinates_scripts/getcoord_pathfind_return.cpp
#include <opencv2/opencv.hpp>
#include <nlohmann/json.hpp>
#include <queue>
#include <unordered_set>
#include <cmath>
#include <limits>

using json = nlohmann::json;

class PathfindingConverter {
private:
    // Custom hash function for pair<int, int>
    struct PairHash {
        template <class T1, class T2>
        std::size_t operator () (const std::pair<T1, T2>& p) const {
            auto h1 = std::hash<T1>{}(p.first);
            auto h2 = std::hash<T2>{}(p.second);
            return h1 ^ h2;
        }
    };

    // Node class for A* pathfinding
    class Node {
    public:
        std::pair<int, int> position;
        Node* parent;
        double g;  // Cost from start to current node
        double h;  // Estimated cost from current to goal
        double f;  // Total cost

        Node(std::pair<int, int> pos, Node* par = nullptr) 
            : position(pos), parent(par), g(0), h(0), f(0) {}

        // Custom comparator for priority queue
        struct Compare {
            bool operator()(const Node* a, const Node* b) const {
                return a->f > b->f;
            }
        };
    };

    // Heuristic function (Euclidean distance)
    static double heuristic(const std::pair<int, int>& a, const std::pair<int, int>& b) {
        int dx = a.first - b.first;
        int dy = a.second - b.second;
        return std::sqrt(dx*dx + dy*dy);
    }

    // Check if a position is walkable
    static bool isWalkable(const cv::Mat& object_map, const std::pair<int, int>& position) {
        int row = position.first;
        int col = position.second;

        // Check bounds
        if (row < 0 || row >= object_map.rows || 
            col < 0 || col >= object_map.cols) {
            return false;
        }

        // Check pixel value
        if (object_map.channels() == 3) {
            // Color image: check if all channels are 255
            cv::Vec3b pixel = object_map.at<cv::Vec3b>(row, col);
            return pixel[0] == 255 && pixel[1] == 255 && pixel[2] == 255;
        } else {
            // Grayscale image: check if pixel is 255
            return object_map.at<uchar>(row, col) == 255;
        }
    }

public:
    static std::string get(const cv::Mat& object_map, 
                           double resolution, 
                           const std::pair<double, double>& origin, 
                           const json& items_data, 
                           const json& robot_coords, 
                           const std::string& assistant_reply) {
        try {
            // Parse assistant reply
            json assistant_data = json::parse(assistant_reply);
            std::string target_id = assistant_data.value("target_id", "");

            // Find target item coordinates
            json item_coords;
            for (const auto& [item_class, items_list] : items_data["items"].items()) {
                for (const auto& item : items_list) {
                    if (item["id"] == target_id) {
                        item_coords = item["coordinates"];
                        break;
                    }
                }
                if (!item_coords.empty()) break;
            }

            if (item_coords.empty()) {
                assistant_data["success"] = "false";
                assistant_data["error"] = "Target ID not found in items_data";
                assistant_data["message"] = "The specified target could not be found.";
                return assistant_data.dump();
            }

            // Map dimensions
            int map_height = object_map.rows;
            int map_width = object_map.cols;

            // Extract origin coordinates
            double origin_x = origin.first;
            double origin_y = origin.second;

            // Convert world to map coordinates
            auto world_to_map = [&](double x, double y) {
                int col = static_cast<int>((x - origin_x) / resolution);
                int row = map_height - static_cast<int>((y - origin_y) / resolution) - 1;
                return std::make_pair(row, col);
            };

            // Convert robot and item positions
            double robot_x = robot_coords["x"];
            double robot_y = robot_coords["y"];
            auto robot_position = world_to_map(robot_x, robot_y);

            double item_x = item_coords["x"];
            double item_y = item_coords["y"];
            auto item_position = world_to_map(item_x, item_y);

            // Check map bounds
            auto is_within_bounds = [&](const std::pair<int, int>& pos) {
                return pos.first >= 0 && pos.first < map_height && 
                       pos.second >= 0 && pos.second < map_width;
            };

            if (!is_within_bounds(robot_position)) {
                assistant_data["success"] = "false";
                assistant_data["error"] = "Robot coordinates out of map bounds";
                assistant_data["message"] = "The robot's position is outside the map boundaries.";
                return assistant_data.dump();
            }

            if (!is_within_bounds(item_position)) {
                assistant_data["success"] = "false";
                assistant_data["error"] = "Item coordinates out of map bounds";
                assistant_data["message"] = "The item's position is outside the map boundaries.";
                return assistant_data.dump();
            }

            // A* Pathfinding
            std::priority_queue<Node*, std::vector<Node*>, Node::Compare> open_list;
            std::unordered_set<std::pair<int, int>, PairHash> closed_set;

            // Movements: up, down, left, right
            std::vector<std::pair<int, int>> movements = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};

            // Starting node
            Node* start_node = new Node(robot_position);
            start_node->h = heuristic(robot_position, item_position);
            start_node->f = start_node->h;

            double min_distance = start_node->h;
            Node* best_node = start_node;

            open_list.push(start_node);

            while (!open_list.empty()) {
                Node* current_node = open_list.top();
                open_list.pop();

                auto current_position = current_node->position;

                // Skip if already processed
                if (closed_set.count(current_position)) continue;

                closed_set.insert(current_position);

                // Update best node if closer to the item
                double current_distance = heuristic(current_position, item_position);
                if (current_distance < min_distance) {
                    min_distance = current_distance;
                    best_node = current_node;
                }

                // If reached the item position, break
                if (current_position == item_position) {
                    best_node = current_node;
                    break;
                }

                // Explore neighbors
                for (const auto& move : movements) {
                    std::pair<int, int> neighbor_position = {
                        current_position.first + move.first, 
                        current_position.second + move.second
                    };

                    if (!isWalkable(object_map, neighbor_position) || 
                        closed_set.count(neighbor_position)) continue;

                    Node* neighbor_node = new Node(neighbor_position, current_node);
                    neighbor_node->g = current_node->g + 1;
                    neighbor_node->h = heuristic(neighbor_position, item_position);
                    neighbor_node->f = neighbor_node->g + neighbor_node->h;

                    open_list.push(neighbor_node);
                }
            }

            // Return map coordinates
            assistant_data["coordinates"] = {
                {"x", best_node->position.second},  // col index as x-coordinate
                {"y", best_node->position.first}    // row index as y-coordinate
            };

            assistant_data["success"] = "true";
            assistant_data["error"] = "none";

            // Clean up dynamically allocated nodes
            std::unordered_set<Node*> to_delete;
            Node* current = best_node;
            while (current) {
                to_delete.insert(current);
                current = current->parent;
            }
            for (auto node : to_delete) {
                delete node;
            }

            return assistant_data.dump();

        } catch (const std::exception& e) {
            json error_response = {
                {"success", "false"},
                {"error", "PathfindingError"},
                {"message", std::string("Error in pathfinding: ") + e.what()}
            };
            return error_response.dump();
        }
    }
};